Introduce contribution guidelines and source documentation

We currently do not have any explicit contribution guidelines regarding common situations like discussing with us over an issue before working on a large feature and other general commandments typically found in open source projects, and it would certainly be nice to have some.

While at it, I'd also like to create a separate document (or set of documents) where we go through and record reasonings and explanations for certain patterns or one-off "hacks" as both an easier way for contributors to get familiar with the code as well as for existing maintainers to be able to document the "implicit knowledge" that needs to be preserved for the next crop of developers who take over this project as and when the current leadership steps away.

Along these lines, I am having a hard time setting up the tooling for development so I can contribute. I have been battling through a variety of errors. Perhaps if we just work through them here, it could later be cleaned up into a SETUP.md or something

My current setup in VS Code on Fedora 32.

Get the code and check SDK version,
```
git clone -b release git@github.com:android-password-store/Android-Password-Store.git
grep compileSdkVersion buildSrc/src/main/java/BaseProjectConfig.kt
#    compileSdkVersion(29)
```

Find most recent commandline tools link https://developer.android.com/studio#downloads and download.

```
unzip commandlinetools-linux-6858069_latest.zip
cd cmdline-tools/

# Accept all licenses and install SDK. Use version from above
yes | ./bin/sdkmanager --sdk_root=. `./bin/sdkmanager --sdk_root=. --list | grep android-29 | cut -f1 -d\|`
```

~/.bashrc
```
export ANDROID_SDK_ROOT=~/.local/cmdline-tools/
export GRADLE_OPTS=-XX:MaxPermSize=1024m
```

At this point, I start getting a variety of build errors when I run `gradle build --stacktrace`. Sometimes I get different errors.

```
FAILURE: Build failed with an exception.

* What went wrong:
Gradle build daemon has been stopped: JVM garbage collector thrashing and after running out of JVM memory

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.launcher.daemon.server.api.DaemonStoppedException: Gradle build daemon has been stopped: JVM garbage collector thrashing and after running out of JVM memory
```

Thanks. I've gotten to this next error with less memory before, but more memory seems faster

```
Fix the issues identified by lint, or add the following to your build script to proceed with errors:
...
android {
    lintOptions {
        abortOnError false
    }
}
...
The first 3 errors (out of 6) were:

/home/aidan/foss/Android-Password-Store/app/src/main/java/com/zeapo/pwdstore/autofill/oreo/AutofillResponseBuilder.kt:33: Error: Call requires API level 26 (current min is 23): formOrigin [NewApi]
    private val formOrigin = form.formOrigin
                                  ~~~~~~~~~~
/home/aidan/foss/Android-Password-Store/app/src/main/java/com/zeapo/pwdstore/autofill/oreo/AutofillResponseBuilder.kt:34: Error: Call requires API level 26 (current min is 23): scenario [NewApi]
    private val scenario = form.scenario
                                ~~~~~~~~
/home/aidan/foss/Android-Password-Store/app/src/main/java/com/zeapo/pwdstore/autofill/oreo/AutofillResponseBuilder.kt:35: Error: Call requires API level 26 (current min is 23): ignoredIds [NewApi]
    private val ignoredIds = form.ignoredIds
                                  ~~~~~~~~~~
```

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -1,0 +1,43 @@
+# Contribution Guidelines
+
+:tada: First off, thanks for taking the time to contribute! :tada:
+
+This document should be treated as suggestions, and not rules, on how to contribute to Android Password Store. All positive contribution is welcome, so please do not hesitate in pitching forth any ideas you have!
+
+## Table of contents
+
+- [Navigating the source code](#navigating-the-source-code)
+- [Building the project](#building-the-project)
+- [Things to do before you start writing code](#things-to-do-before-you-start-writing-code)
+
+## Navigating the source code
+
+The source code is split across 4 modules.
+
+- `buildSrc` hosts the Gradle build logic for the project
+- `autofill-parser` is the aptly named Autofill logic parser
+- `openpgp-ktx` contains the glue code that enables APS to interact with OpenKeychain
+- `app` is everything else that constitutes APS
+
+In most scenarios, the `app` directory is where you'd be contributing changes to. While most of the code has been rewritten and documented, there are still gnarly "legacy" parts that might be challenging to understand at a glance. Please get in touch via the [Discussions](https://github.com/android-password-store/Android-Password-Store/discussions) page with any questions you have, and we'd love to explain and improve things.
+
+## Building the project
+
+### Building with Gradle
+
+This document assumes that you already have an Android development environment ready. If not, refer to Google's documentation on [installing Android Studio](https://developer.android.com/studio/install). APS will always build against the latest stable release of Android Studio, but you can use pre-release versions of the IDE should you desire so.
+
+The app comes in two 'flavors', a FOSS-only **free** variant and a **nonFree** variant that contains proprietary Google dependencies to facilitate some additional features as documented [here](https://android-password-store.github.io/docs/users/build-types). Decide what flavor you want to build, then run the following command to generate a debug APK.
+
+```shell
+./gradlew assembleFreeDebug # for 'free' flavor
+./gradlew assembleNonFreeDebug # for 'nonFree' flavor
+```
+
+You can find the generated APK at `app/build/outputs/apk/<variant>/debug/app-<variant>-debug.apk`.
+
+# Things to do before you start writing code
+
+If you're trying to fix a bug that already has an open issue, it's a good idea to drop a comment mentioning that you're working on a fix. If no open issue exists, ensure that you explain the bug you're fixing in some detail in the pull request body. This helps us as maintainers get a better sense of why you're making specific changes, and we might have insight into better ways of fixing the problem.
+
+If you want to add a new feature, please file an issue first to gauge maintainer and community interest. Sometimes the feature might need adjustments before it is accepted, or it might not be something we want to add to APS at all. Proposing the feature in detail before working on it ensures that you build exactly what is necessary and not waste time and effort on aspects that would get rejected during review.

